File: packages/python/src/bharat_choropleth/topojson.py
"""Small TopoJSON decoder used by the static renderers.

It supports the area geometries a choropleth needs: ``Polygon`` and
``MultiPolygon``.  Shared arcs, reversed arc indexes, quantized/delta encoded
coordinates, holes, and disjoint islands are retained.  No third-party
TopoJSON package is required.
"""

from __future__ import annotations

from dataclasses import dataclass
import json
from os import PathLike
from pathlib import Path
from typing import Any, Dict, Iterable, List, Mapping, Optional, Sequence, Tuple, Union

Point = Tuple[float, float]
Ring = Tuple[Point, ...]
TopologyInput = Union[Mapping[str, Any], str, PathLike[str]]


class TopoJSONError(ValueError):
    """Raised when a supplied TopoJSON mapping cannot be decoded."""


@dataclass(frozen=True)
class Feature:
    """One decoded polygon feature.

    ``rings`` is intentionally flat.  SVG's ``fill-rule=evenodd`` correctly
    renders both holes and every disjoint island without requiring callers to
    classify rings.  Coordinates are ``(longitude, latitude)`` pairs.
    """

    id: str
    name: str
    rings: Tuple[Ring, ...]
    properties: Mapping[str, Any]


def load_topology(source: TopologyInput) -> Mapping[str, Any]:
    """Load a TopoJSON mapping from a mapping or a UTF-8 JSON file path."""

    if isinstance(source, Mapping):
        return source
    try:
        with Path(source).open("r", encoding="utf-8") as handle:
            decoded = json.load(handle)
    except (OSError, TypeError, json.JSONDecodeError) as error:
        raise TopoJSONError("could not load TopoJSON input") from error
    if not isinstance(decoded, Mapping):
        raise TopoJSONError("TopoJSON document must be a JSON object")
    return decoded


def decode_topology(
    source: TopologyInput,
    *,
    object_name: Optional[str] = None,
    id_property: str = "id",
    name_property: str = "name",
) -> Tuple[Feature, ...]:
    """Decode a TopoJSON object to area features.

    ``object_name`` selects an entry in the top-level ``objects`` mapping.  If
    omitted, the first object is decoded, matching the generated bundles in
    this repository.  A geometry id wins over ``id_property`` only when that
    property is absent, allowing stable caller-owned ids.
    """

    topology = load_topology(source)
    objects = topology.get("objects")
    if not isinstance(objects, Mapping) or not objects:
        raise TopoJSONError('topology has no non-empty "objects" mapping')
    selected_name = object_name or next(iter(objects))
    selected = objects.get(selected_name)
    if not isinstance(selected, Mapping):
        raise TopoJSONError('topology has no object named "{}"'.format(selected_name))

    arcs = _decode_arcs(topology)
    geometries: Iterable[Any]
    if selected.get("type") == "GeometryCollection":
        geometries = selected.get("geometries", ())
    else:
        geometries = (selected,)

    features: List[Feature] = []
    for geometry in geometries:
        if not isinstance(geometry, Mapping):
            continue
        rings = _geometry_rings(geometry, arcs)
        if rings is None:
            continue
        raw_properties = geometry.get("properties", {})
        properties: Mapping[str, Any] = raw_properties if isinstance(raw_properties, Mapping) else {}
        raw_id = properties.get(id_property, geometry.get("id", properties.get(name_property, "")))
        feature_id = str(raw_id)
        name = str(properties.get(name_property, feature_id))
        features.append(Feature(feature_id, name, rings, properties))
    return tuple(features)


def _decode_arcs(topology: Mapping[str, Any]) -> Tuple[Ring, ...]:
    raw_arcs = topology.get("arcs")
    if not isinstance(raw_arcs, Sequence) or isinstance(raw_arcs, (str, bytes)):
        raise TopoJSONError('topology has no "arcs" list')
    transform = topology.get("transform")
    scale: Optional[Tuple[float, float]] = None
    translate: Optional[Tuple[float, float]] = None
    if transform is not None:
        if not isinstance(transform, Mapping):
            raise TopoJSONError("transform must be an object")
        scale = _pair(transform.get("scale"), "transform.scale")
        translate = _pair(transform.get("translate"), "transform.translate")

    arcs: List[Ring] = []
    for raw_arc in raw_arcs:
        if not isinstance(raw_arc, Sequence) or isinstance(raw_arc, (str, bytes)):
            raise TopoJSONError("an arc is not a list of positions")
        x = 0.0
        y = 0.0
        points: List[Point] = []
        for raw_position in raw_arc:
            px, py = _pair(raw_position, "arc position")
            if scale is not None and translate is not None:
                x += px
                y += py
                points.append((x * scale[0] + translate[0], y * scale[1] + translate[1]))
            else:
                points.append((px, py))
        arcs.append(tuple(points))
    return tuple(arcs)


def _pair(value: Any, label: str) -> Tuple[float, float]:
    if not isinstance(value, Sequence) or isinstance(value, (str, bytes)) or len(value) < 2:
        raise TopoJSONError("{} must be a two-number list".format(label))
    try:
        return float(value[0]), float(value[1])
    except (TypeError, ValueError) as error:
        raise TopoJSONError("{} must contain numbers".format(label)) from error


def _geometry_rings(geometry: Mapping[str, Any], arcs: Tuple[Ring, ...]) -> Optional[Tuple[Ring, ...]]:
    geometry_type = geometry.get("type")
    raw_indexes = geometry.get("arcs", ())
    if geometry_type not in ("Polygon", "MultiPolygon"):
        return None
    if not isinstance(raw_indexes, Sequence) or isinstance(raw_indexes, (str, bytes)):
        return tuple()
    rings: List[Ring] = []
    if geometry_type == "Polygon":
        for ring_indexes in raw_indexes:
            rings.append(_stitch(ring_indexes, arcs))
    else:
        for polygon in raw_indexes:
            if not isinstance(polygon, Sequence) or isinstance(polygon, (str, bytes)):
                raise TopoJSONError("a multipolygon part is not a list of rings")
            for ring_indexes in polygon:
                rings.append(_stitch(ring_indexes, arcs))
    return tuple(rings)


def _stitch(raw_indexes: Any, arcs: Tuple[Ring, ...]) -> Ring:
    if not isinstance(raw_indexes, Sequence) or isinstance(raw_indexes, (str, bytes)):
        raise TopoJSONError("a ring is not a list of arc indexes")
    stitched: List[Point] = []
    for raw_index in raw_indexes:
        if not isinstance(raw_index, (int, float)):
            raise TopoJSONError("an arc index is not numeric")
        index = int(raw_index)
        arc_index = ~index if index < 0 else index
        if arc_index < 0 or arc_index >= len(arcs):
            raise TopoJSONError("arc index {} is out of range".format(index))
        arc = arcs[arc_index]
        points = tuple(reversed(arc)) if index < 0 else arc
        stitched.extend(points if not stitched else points[1:])
    return tuple(stitched)

File: packages/flutter/lib/src/topojson.dart
import 'dart:ui' show Offset;

/// Thrown when a payload isn't TopoJSON this decoder can read.
class TopoJsonException implements Exception {
  TopoJsonException(this.message);
  final String message;
  @override
  String toString() => 'TopoJsonException: $message';
}

/// One map region's geometry and identity, decoded from TopoJSON.
///
/// [rings] holds every closed ring the feature is made of — outer rings *and*
/// holes, and every part of a multi-part feature, flattened into one list.
/// That's deliberate: the renderer fills with [PathFillType.evenOdd], so holes
/// (an enclave inside a state) and islands (Andaman & Nicobar) both come out
/// right without the caller tracking which ring is which.
class MapFeature {
  const MapFeature({
    required this.id,
    required this.name,
    required this.rings,
    this.properties = const {},
  });

  /// Stable identifier — LGD-derived (`in-cs-30-goa`) in the bundled data.
  final String id;

  /// Display name (`Goa`).
  final String name;

  /// Every ring, in lon/lat degrees. `Offset.dx` is longitude, `dy` is latitude.
  final List<List<Offset>> rings;

  final Map<String, Object?> properties;

  @override
  String toString() => 'MapFeature($id, $name, ${rings.length} rings)';
}

/// Decode a TopoJSON topology into features.
///
/// [objectName] picks which named object to unpack (`states`, `districts`).
/// When omitted the first object is used, which is what the prepared bundles in
/// this repository have.
///
/// Only Polygon and MultiPolygon geometries are read; anything else is skipped,
/// since a choropleth has nothing to fill for points and lines.
List<MapFeature> decodeTopoJson(
  Map<String, Object?> topology, {
  String? objectName,
  String Function(Map<String, Object?> properties)? getId,
  String Function(Map<String, Object?> properties)? getName,
}) {
  final objects = topology['objects'];
  if (objects is! Map<String, Object?> || objects.isEmpty) {
    throw TopoJsonException('topology has no "objects" — is this TopoJSON?');
  }

  final key = objectName ?? objects.keys.first;
  final object = objects[key];
  if (object == null) {
    throw TopoJsonException('no object named "$key" (have: ${objects.keys.join(", ")})');
  }
  if (object is! Map<String, Object?>) {
    throw TopoJsonException('object "$key" is not a TopoJSON geometry object');
  }

  final arcs = _decodeArcs(topology);

  final geometries = object['type'] == 'GeometryCollection'
      ? (object['geometries'] as List? ?? const [])
      : [object];

  final features = <MapFeature>[];
  for (final geometry in geometries) {
    if (geometry is! Map<String, Object?>) continue;
    // Null means "not an area geometry" and is skipped. An *empty* ring list is
    // different: it's an area feature the source has no parts for (Lakshadweep
    // in the bundled current-2019 states). Those are kept, so ids, labels and
    // values still line up with the same data rendered on the web — the region
    // simply draws nothing.
    final rings = _ringsOf(geometry, arcs);
    if (rings == null) continue;

    final properties = (geometry['properties'] as Map?)?.cast<String, Object?>() ?? const <String, Object?>{};
    final id = getId?.call(properties) ??
        (properties['id'] ?? geometry['id'] ?? properties['name'] ?? '').toString();
    final name = getName?.call(properties) ?? (properties['name'] ?? id).toString();

    features.add(MapFeature(id: id, name: name, rings: rings, properties: properties));
  }
  return features;
}

/// Arc positions are delta-encoded when the topology is quantized: each point is
/// stored as an offset from the previous one, and a shared transform maps the
/// running total back into lon/lat. Un-quantized topologies carry absolute
/// coordinates and no transform.
List<List<Offset>> _decodeArcs(Map<String, Object?> topology) {
  final rawArcs = topology['arcs'];
  if (rawArcs is! List) throw TopoJsonException('topology has no "arcs"');

  final transform = topology['transform'];
  List<double>? scale;
  List<double>? translate;
  if (transform is Map) {
    scale = _pair(transform['scale'], 'transform.scale');
    translate = _pair(transform['translate'], 'transform.translate');
  }

  return rawArcs.map<List<Offset>>((arc) {
    if (arc is! List) throw TopoJsonException('an arc is not a list of positions');
    var x = 0.0;
    var y = 0.0;
    final points = <Offset>[];
    for (final position in arc) {
      if (position is! List || position.length < 2) {
        throw TopoJsonException('an arc position is not a [x, y] pair');
      }
      final px = (position[0] as num).toDouble();
      final py = (position[1] as num).toDouble();
      if (scale != null && translate != null) {
        x += px;
        y += py;
        points.add(Offset(x * scale[0] + translate[0], y * scale[1] + translate[1]));
      } else {
        points.add(Offset(px, py));
      }
    }
    return points;
  }).toList(growable: false);
}

List<double> _pair(Object? value, String label) {
  if (value is! List || value.length < 2) {
    throw TopoJsonException('$label must be a two-number list');
  }
  return [(value[0] as num).toDouble(), (value[1] as num).toDouble()];
}

/// Returns null for geometry a choropleth cannot fill (points, lines), and a
/// possibly-empty ring list for Polygon/MultiPolygon.
List<List<Offset>>? _ringsOf(Map<String, Object?> geometry, List<List<Offset>> arcs) {
  final type = geometry['type'];
  if (type != 'Polygon' && type != 'MultiPolygon') return null;

  final arcIndices = geometry['arcs'];
  if (arcIndices is! List) return const [];

  if (type == 'Polygon') {
    // arcs: [ring, ...] where ring is [arcIndex, ...]
    return arcIndices.map((ring) => _stitch(ring as List, arcs)).toList();
  }
  // arcs: [polygon, ...] where polygon is [ring, ...]
  return [
    for (final polygon in arcIndices)
      for (final ring in polygon as List) _stitch(ring as List, arcs),
  ];
}

/// Join a ring's arcs end to end. A negative index means "this arc, reversed" —
/// `~index` is the arc it refers to, matching the TopoJSON spec. Consecutive
/// arcs share their touching endpoint, so every arc after the first drops its
/// leading point to avoid duplicating it.
List<Offset> _stitch(List<Object?> arcIndices, List<List<Offset>> arcs) {
  final ring = <Offset>[];
  for (final raw in arcIndices) {
    final index = (raw as num).toInt();
    final reversed = index < 0;
    final arcIndex = reversed ? ~index : index;
    if (arcIndex < 0 || arcIndex >= arcs.length) {
      throw TopoJsonException('arc index $index is out of range');
    }
    final arc = arcs[arcIndex];
    final points = reversed ? arc.reversed : arc;
    ring.addAll(ring.isEmpty ? points : points.skip(1));
  }
  return ring;
}
